packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -1302,6 +1303,82 @@ export class MergeTree {
 		return undefined;
 	}
 
+	/**
+	 * Finds the nearest reference with ReferenceType.Tile to `startPos` in the direction dictated by `posPrecedesTile`.
+	 * Uses depthFirstNodeWalk in addition to the block-accelerated functionality introduced for findTile. The search
+	 * position will be included in the nodes to walk, so searching on all positions, including the endpoints, can be considered
+	 * inclusive. Any out of bound search positions will return undefined, so in order to search the whole string, a forward
+	 * search can begin at 0, or a bacward search can begin at length-1.This function will replace findTile, which will be deprecated.
+	 *
+	 * @param startPos - Position at which to start the search
+	 * @param clientId - clientId dictating the perspective to search from
+	 * @param tileLabel - Label of the tile to search for
+	 * @param tilePrecedesPos - Whether the desired tile comes before (true) or after (false) `startPos`
+	 */
+	public walkToFindTile(
+		startPos: number,
+		clientId: number,
+		tileLabel: string,
+		tilePrecedesPos = true,
+	) {
+		const searchInfo: IReferenceSearchInfo = {
+			mergeTree: this,
+			tilePrecedesPos,
+			tileLabel,
+		};
+
+		const tileShiftHelper = (node: IMergeNode) => {
+			tileShift(node, 0, UniversalSequenceNumber, clientId, startPos, -1, searchInfo);
```

Comment:
This or some other helper should sometimes be returning `NodeAction.Skip` in order to achieve the former method's log(N) complexity. As-is you probably have a method which walks all nodes, and we can do better.

I think since `depthFirstNodeWalk` assumes you already have the segment+offset, you'll actually need to use the up action rather than the down action (or maybe both). Starting at a segment + its parent, you want to iterate forward over each segment looking for markers, then over each subsequent segment-parent block, then each segment-grandparent block, etc. until you find a block with a leftmost tile (when searching forward) or rightmost tile (when searching backward)

Reply:
here's a picture of what I mean if merge-tree's branching factor were 3 and not 8.

![image](https://github.com/microsoft/FluidFramework/assets/5222607/8752c172-17aa-4177-a07f-83a97aefb0cc)

Red is the segment that the position resolves to. Blue are markers. Orange is the path of the nodes I'd expect to visit.